### PR TITLE
V5

### PR DESCRIPTION
**Pull request by danielbarion**
[Upgrade Guide in new docs](https://react-tooltip.com/docs/category/upgrade-guide)

## Installing

```
npm install react-tooltip@5.0.0-beta.2
```
or
```
yarn add react-tooltip@5.0.0-beta.2
```

### Current exported files
image

# Changelog V4 -> V5

## Improvements

- Dropped package dependency `uuid`
- - Using React `useId` (https://beta.reactjs.org/apis/react/useId)
- - - Unfortunately `useId` was introduced only into React v18, so, that will be the minimum necessary version of React to V5
- Dropped package dependency `prop-types` (I'm thinking if we will use prop-types with TS)

## `V4` props available in `V5`:

- [x] children - 
- [x] place - `data-place`
- [x] ~type~ - Deprecated | in V5 -> `variant` - `data-variant`
- [ ] effect - not implemented yet, if many users need this feature, we will work on this one.
- [x] offset - `data-offset`
- [ ] ~padding~ - Deprecated | in V5 -> can be easy updated by className prop
- [ ] ~multiline~ - Deprecated | in V5 -> this is already supported as default by `content` and `html` props
- [ ] ~border~ - Deprecated | in V5 -> can be easy updated by `className` prop
- [ ] ~borderClass~ - Deprecated | in V5 -> can be easy updated by `className` prop
- [ ] ~textColor~ - Deprecated | in V5 -> can be easy updated by `className` prop
- [ ] ~backgroundColor~ - Deprecated | in V5 -> can be easy updated by `className` prop
- [ ] ~borderColor~ - Deprecated | in V5 -> can be easy updated by `className` prop
- [ ] ~arrowColor~ - Deprecated | in V5 -> can be easy updated by `className` prop
- [ ] ~arrowRadius~ - Deprecated | in V5 -> can be easy updated by `className` prop
- [ ] ~tooltipRadius~ - Deprecated | in V5 -> can be easy updated by `className` prop
- [ ] ~insecure~ - Deprecated | in V5 -> CSS will be a separate file and can be imported or not
- [x] className
- [x] id
- [x] html
- [x] delayHide - `data-delay-hide`
- [ ] ~delayUpdate~ - Deprecated | if requested, can be implemented later
- [x] delayShow - `data-delay-show`
- [ ] event - not implemented yet, if many users need this feature, we will work on this one.
- [ ] ~eventOff~ - Deprecated
- [ ] ~isCapture~ - Deprecated
- [ ] ~globalEventOff~ - Deprecated
- [x] getContent - Now this attribute only accepts a function instead of any
- [ ] afterShow - not implemented yet, if many users need this feature, we will work on this one.
- [ ] afterHide - not implemented yet, if many users need this feature, we will work on this one.
- [ ] ~overridePosition~ - Deprecated
- [ ] ~disable~ - Deprecated | in V5 -> state can be controlled or uncontrolled
- [ ] scrollHide - not implemented yet, if many users need this feature, we will validate if we wrok on this one.
- [ ] resizeHide - not implemented yet, if many users need this feature, we will validate if we wrok on this one.
- [x] wrapper - `data-wrapper`
- [ ] ~bodyMode~ - Deprecated
- [ ] ~clickable~ - Deprecated | Supported by default in V5
- [ ] ~disableInternalStyle~ - Deprecated | in V5 -> CSS will be a separate file and can be imported or not

## Break Changes

- All data attributes now has `tooltip` into his name
- Default Padding changed from `padding: 8px 21px;` to `padding: 8px 16px;`
- Exported module now is `Tooltip` instead of `ReactTooltip`
- - If you already have a `Tooltip` component in your application and want to explicitly declare this is `ReactTooltip`, just `import { Tooltip as ReactTooltip } from "react-tooltip"`
- CSS import is now optional, so, you can modify and/or add any styling to your floating tooltip element
- `data-tip` attribute now is `data-content`
- `getContent` prop now doesn't accept `any` anymore, just a `function`
- default behavior of tooltip now is `solid` instead of `float`

## New Props

- [x] classNameArrow
- [x] events - `data-events` -`['hover', 'click']` - default: `['hover']` (always an array when using as prop, even with only one option, when using as data attribute: `data-events="hover click"`)
- [x] isOpen - `boolean` (to control tooltip state) - if not used, internal state of tooltip will handle the show state
- [x] setIsOpen - `function` (to control tooltip state) - if not used, internal state of tooltip will handle the show state

### CODE REVIEW

**Comment by danielbarion** [comment]
Hi guys, I want to know your thoughts about a thing, as in the current code of V5, we only have 2 elements for tooltip:
- tooltip div 
- arrow div
And there are a lot of props related to tooltip customization and arrow customization, like:
- border
- borderClass
- padding
-  backgroundColor
- borderColor
- arrowColor
- arrowRadius
- tooltipRadius

And we can override the default styles by simply using CSS in prop `className` and `classNameArrow`.

So, those v4 props sound redundant to me, what does everyone thinks about that?

**Comment by aronhelser** [comment]
I think you should make it as simple as possible, still keeping the same customization available. So if you are able to make the changes with pure CSS, it is fine to leave these props out - but you will want to make a migration guide + example to show how to change them with CSS in the new version.

It's great you are able to work on this and possibly get to a more react style component with the new version! I'm sure many people are looking forward to it!

**Comment by danielbarion** [comment]
hey @aronhelser, I'm very happy to have your feedback, thanks!

Yes, sure! I'll update docs and the release notes with the steps and changes after every change :)

**Comment by FedeTommi** [comment]
About [#753](https://github.com/ReactTooltip/react-tooltip/issues/753) : it would be easily implemented if the size of the arrow was the same in every orientation. Currently there are some pixels of difference that would make the calculations harder.

**Comment by danielbarion** [comment]
Yes, I agree @FedeTommi, the current code for arrow in V5 is:

```
.arrow {
  position: absolute;
  background: #222;
  width: 8px;
  height: 8px;
  transform: rotate(45deg);
}
```

So, probably this problem will not happen in V5, as we are using the API of `float-ui` to compute the positions.

File: .gitignore [hunk 1]
```diff
@@ -1,16 +1,37 @@
...
-*.log
-.DS_Store*
+## editors
+/.idea
+
+## system files
+.DS_Store
 .idea/
+*.log
+
```
Comment (erezcarmel) [inline .gitignore:8]: As VSCode becomes more and more popular IDE, I'd recommend adding its files to the `.gitignore` list
```suggestion
.vscode/*
!.vscode/settings.json
!.vscode/tasks.json
!.vscode/launch.json
!.vscode/extensions.json
```

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (erezcarmel) [inline .vscode/settings.json:1]: VSCode IDE setting file is a personal flavor, and better not be committed.

File: src/components/Tooltip/styles.module.css [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  height: 8px;
+  transform: rotate(45deg);
+}
+
+.show {
+  visibility: visible;
+  opacity: 0.9;
+}
+
+/** Types variant **/
```
Comment (erezcarmel) [inline src/components/Tooltip/styles.module.css:29]: I'd recommend moving all default color values to CSS constants in default styles CSS file:
```
:root {
  --dark-background: #222;
  --dark-text-color: white;

...
}
```

and use it:
```
background: var(--dark-background);
color: var(--dark-text-color);

...
```

This way there will be one place for color changes and adding variant colors in the future

File: .vscode/settings.json (hunk 2 above)
Comment (danielbarion) [reply .vscode/settings.json:1]: I agree, thanks!

File: src/components/Tooltip/styles.module.css (hunk 3 above)
Comment (danielbarion) [reply src/components/Tooltip/styles.module.css:29]: Great!

I like CSS vars, thanks for the suggestion!

File: package-old.json [hunk 4]
```diff
@@ -0,0 +1,175 @@
+{
```
Comment (danielbarion) [inline package-old.json:1]: This file will be deleted when the PR is ready, I'm just letting this file here to compare and double check if everything is ok

**Comment by danielbarion** [comment]
This is WIP but it's already available for preview:

https://react-tooltip.com/docs/getting-started

**Comment by danielbarion** [comment]
Looking forward to have feedback on the work that is already done until now

**Comment by danielbarion** [comment]
`v5.0.0-beta.1` released to fix types export and now a typescript project can use the react tooltip as expected :)

**Comment by danielbarion** [comment]
Hi @aronhelser can you let me know your thoughts on https://react-tooltip.com/docs/getting-started, please?
Thanks!

**Comment by aronhelser** [comment]
It looks great! I don't have particular comments about code style or anything. I see a few typos various places in docs/readme, but I don't think that should hold you up. I haven't worked in typescript in 7 years, so I would defer to someone else there.

+1

**Comment by danielbarion** [comment]
> It looks great! I don't have particular comments about code style or anything. I see a few typos various places in docs/readme, but I don't think that should hold you up. I haven't worked in typescript in 7 years, so I would defer to someone else there.
> 
> +1

Thanks :D 

oh, the typos are in Grammarly?

Please, do the suggestions, I'm not a native English speaker, but I'm still improving my English hahaha

**Comment by danielbarion** [comment]
I'll release ReactTooltip V5 tomorrow (December 7), if anyone has anything to say before that, now it's the moment :)

**Comment by danielbarion** [comment]
All issues related to V4 will be closed.

They are old and the codebase from V4 to V5 is totally different.

**Review by erezcarmel — APPROVED** [review]
Looks great! goodluck to V5!

**Pull request merged**